service/src/test/java/bio/terra/workspace/service/resource/controlled/flight/clone/workspace/CloneAllFoldersStepTest.java

```diff
@@ -0,0 +1,134 @@
+
+  @Test
+  public void doStep_foldersCloned() throws InterruptedException {
+    var workingMap = new FlightMap();
+    var inputParameters = new FlightMap();
+
+    UUID destinationWorkspaceId =
+        WorkspaceUnitTestUtils.createWorkspaceWithGcpContext(workspaceDao);
+    Workspace destinationWorkspace = workspaceDao.getWorkspace(destinationWorkspaceId);
+
+    inputParameters.put(ControlledResourceKeys.SOURCE_WORKSPACE_ID, SOURCE_WORKSPACE_ID);
+    inputParameters.put(JobMapKeys.REQUEST.getKeyName(), destinationWorkspace);
+
+    when(mockFlightContext.getInputParameters()).thenReturn(inputParameters);
+    when(mockFlightContext.getWorkingMap()).thenReturn(workingMap);
+
+    StepResult stepResult = cloneAllFoldersStep.doStep(mockFlightContext);
+
+    assertEquals(StepResult.getStepResultSuccess(), stepResult);
+
+    assertEquals(
+        2,
+        folderDao.listFoldersInWorkspace(destinationWorkspaceId).size(),
+        "Destination workspace clones the folders successfully");
+    assertNotNull(
+        folderDao.listFoldersInWorkspace(destinationWorkspaceId).stream()
+            .filter(folder -> folder.displayName().equals(SOURCE_PARENT_FOLDER_NAME)),
+        "Destination parent folder is cloned successfully");
+    Folder destinationParentFolder =
+        folderDao.listFoldersInWorkspace(destinationWorkspaceId).stream()
+            .filter(folder -> folder.displayName().equals(SOURCE_PARENT_FOLDER_NAME))
+            .findFirst()
+            .get();
+    assertNotNull(
+        folderDao.listFoldersInWorkspace(destinationWorkspaceId).stream()
+            .filter(folder -> folder.displayName().equals(SOURCE_SON_FOLDER_NAME))
+            .findFirst(),
+        "Destination son folder is cloned successfully");
+    Folder destinationSonFolder =
+        folderDao.listFoldersInWorkspace(destinationWorkspaceId).stream()
+            .filter(folder -> folder.displayName().equals(SOURCE_SON_FOLDER_NAME))
+            .findFirst()
+            .get();
+
+    assertNotEquals(
+        destinationParentFolder.id(),
+        SOURCE_PARENT_FOLDER_ID,
+        "Destination parent folder id is generated successfully");
+    assertNotEquals(
+        destinationSonFolder.id(),
+        SOURCE_SON_FOLDER_ID,
+        "Destination son folder id is generated successfully");
+    assertEquals(
+        destinationParentFolder.description(),
+        SOURCE_PARENT_FOLDER_DESCRIPTION,
+        "Destination parent folder description is cloned successfully");
+    assertEquals(
+        destinationSonFolder.description(),
+        SOURCE_SON_FOLDER_DESCRIPTION,
+        "Destination son folder description is cloned successfully");
+    assertNull(
+        destinationParentFolder.parentFolderId(),
+        "Destination parent folder parent id cloned successfully");
+    assertEquals(
+        destinationParentFolder.id(),
+        destinationSonFolder.parentFolderId(),
+        "Destination son folder parent id cloned successfully");
+  }
```

Comment:
can you add an undostep test in which you call doStep and then undoStep and assert that the destination workspace has no folder?